File: content/02-two-pointers/valid_palindrome.js
/**
 * Problem: Valid Palindrome
 * LeetCode: #125 (Easy) | NeetCode 150 | Blind 75
 */

function isPalindrome(s) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    let left = 0, right = s.length - 1;
    const isAlphanumeric = (char) => {
        const code = char.charCodeAt(0);
        return (code >= 48 && code <= 57) || // numeric (0-9)
               (code >= 65 && code <= 90) || // upper alpha (A-Z)
               (code >= 97 && code <= 122);  // lower alpha (a-z)
    };

    while (left < right) {
        while (left < right && !isAlphanumeric(s[left])) {
            left++;
        }
        while (left < right && !isAlphanumeric(s[right])) {
            right--;
        }
        if (s[left].toLowerCase() !== s[right].toLowerCase()) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(isPalindrome("A man, a plan, a canal: Panama"), true);
    assert.strictEqual(isPalindrome("race a car"), false);
    assert.strictEqual(isPalindrome(" "), true);
    console.log("JS: isPalindrome passed tests!");
}

File: content/02-two-pointers/valid_palindrome.py
"""
Problem: Valid Palindrome
LeetCode: #125 (Easy) | NeetCode 150 | Blind 75
"""

def is_palindrome(s: str) -> bool:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    left, right = 0, len(s) - 1
    
    while left < right:
        while left < right and not s[left].isalnum():
            left += 1
        while left < right and not s[right].isalnum():
            right -= 1
            
        if s[left].lower() != s[right].lower():
            return False
            
        left += 1
        right -= 1
        
    return True

if __name__ == "__main__":
    assert is_palindrome("A man, a plan, a canal: Panama") == True
    assert is_palindrome("race a car") == False
    assert is_palindrome(" ") == True
    print("Python: is_palindrome passed tests!")
